shaper.go

```
@@ -1,6 +1,141 @@
+	if x >= 0 {
+		x += fixed.Int26_6(unitsPerEm) / 2
+	} else {
+		x -= fixed.Int26_6(unitsPerEm) / 2
+	}
+	return x / fixed.Int26_6(unitsPerEm)
+}
+
+// scalePosition using specific ppem and upem values.
+func scalePosition(gp harfbuzz.GlyphPosition, ppem fixed.Int26_6, upem sfnt.Units) harfbuzz.GlyphPosition {
+	gp.XAdvance = int32(scale(fixed.I(int(gp.XAdvance)).Mul(ppem), upem).Round())
+	gp.YAdvance = int32(scale(fixed.I(int(gp.YAdvance)).Mul(ppem), upem).Round())
+	gp.XOffset = int32(scale(fixed.I(int(gp.XOffset)).Mul(ppem), upem).Round())
+	gp.YOffset = int32(scale(fixed.I(int(gp.YOffset)).Mul(ppem), upem).Round())
+	return gp
+}
+
+// TODO: can we just make Input and Output into structs?
+// Interfaces encapsulate varying behavior well, but these
+// types have no behavior. They're just structured data.
+// It feels silly to wrap them in accessors like this.
+type output struct {
+	DotAdvance    fixed.Int26_6
+	TopToBaseline fixed.Int26_6
+	TextBounds    fixed.Rectangle26_6
+	Glyphs        []Glyph
+}
+
+func (o output) Advance() fixed.Int26_6 {
+	return o.DotAdvance
+}
+
+func (o output) Baseline() fixed.Int26_6 {
+	return o.TopToBaseline
+}
+
+func (o output) Bounds() fixed.Rectangle26_6 {
+	return o.TextBounds
+}
+
+func (o output) Length() int {
+	return len(o.Glyphs)
+}
+
+func (o output) Index(i int) Glyph {
+	return o.Glyphs[i]
+}
+
+// Shape turns an input into an output.
+func Shape(input Input) Output {
+	// Prepare to shape the text.
+	// TODO: maybe reuse these buffers for performance?
+	buf := harfbuzz.NewBuffer()
+	runes, start, end := input.Text()
+	buf.AddRunes(runes, start, end-start)
+	// TODO: handle vertical text?
+	switch input.Direction() {
+	case di.DirectionLTR:
+		buf.Props.Direction = harfbuzz.LeftToRight
+	case di.DirectionRTL:
+		buf.Props.Direction = harfbuzz.RightToLeft
+	}
+	buf.Props.Language = input.Language()
+	buf.Props.Script = input.Script()
+	// TODO: figure out what (if anything) to do if this type assertion fails.
```

I see a couple of options here, we could either return an error, or make input.Face return `harfbuzz.Face` if that is the only thing we support, or we should move `Shape` into more specific packages, eg create a `harfbuzz` package with `Shape(Input) (Output, error)` where those `Input`/`Output` extend a base version, which could be accompanied by an `image` package also containing a `Shape` function but which uses `golang.org/x/image/font.Face`

---

As I do not currently anticipate us using anything other than a `harfbuzz.Face`, I'd be fine changing the type of `input.Face`. Your other option is viable, but seems like overkill. I'm not aware of a use-case for the image/font shaping if harfbuzz shaping is an option.